pallet/common/src/lib.rs

```
@@ -651,6 +651,41 @@ pub trait SFTExt {
 	fn token_exists(token_id: TokenId) -> bool;
 }
 
+/// Trait for creating futurepass accounts
+pub trait FuturepassProvider {
+	type AccountId;
+
+	/// Create a futurepass account for the given owner
+	/// Returns the futurepass account ID on success
+	fn create_futurepass(
+		funder: Self::AccountId,
+		owner: Self::AccountId,
+	) -> Result<Self::AccountId, DispatchError>;
+}
+
+#[cfg(feature = "std")]
+impl FuturepassProvider for () {
+	type AccountId = AccountId;
+
+	fn create_futurepass(
+		_funder: Self::AccountId,
+		owner: Self::AccountId,
+	) -> Result<Self::AccountId, DispatchError> {
+		// Create a deterministic account by hashing the owner's address with a prefix
+		let mut input = Vec::with_capacity(24);
+		// Use a fixed prefix for futurepass accounts (first 4 bytes)
+		input.extend_from_slice(&[0xff, 0xff, 0xff, 0xff]);
+		// Add the owner's account bytes
+		input.extend_from_slice(&owner.encode());
+
+		// Hash the input to get a deterministic address
+		let hash = sp_io::hashing::blake2_256(&input);
+		let address = H160::from_slice(&hash[0..20]);
+
+		Ok(address.into())
+	}
+}
```

nit, if wanting to create some mock behaviour for this trait, it should be done within the mock file under a MockFuturepassProvider instead of as an implementation on the unit type

If wanting to create a unit type impl, this should return an error to prevent accidental behaviour

Personally I think this impl should be removed in favour of using the real Futurepass pallet in the tests to ensure the Futurepass account is created correctly with appropriate values

---

Have mixed opinions on this since - again you create a dependency on the futurepass pallet.
And again the FP pallet has dependencies on other things - mainly the runtime atm.

I have seen many times in the core substrate libs where the unit type is used alongside the trait declaration - I think this is good design as in most cases consumers just want to plug some default impl.
We are providing the same here - in most cases, if users want an impl which wants to hook into storage - then they will need an explicit import of the FP pallet itself as the provider.

The current impl does give a FP address - just without a storage.

---

In this case we should move this to mock. I've seen the unit type impls and have used them myself, however all the impls on the unit type are intended to be no-op, whereas this impl creates some random account from the owner address which.
Though unlikely, if this unit type is accidentally used incorrectly, it could result in unwanted behaviour

---

We have similar implementations in our code where the unit type having mocked-like implementation.

---

No worries, it's not a biggie. Don't want to hold up release for this

---

https://github.com/futureversecom/trn-seed/pull/926/commits/dc5112e1d331092d9782fe9cbfe618e1b1325ac5